docs/contributing/coding-guidelines/languages/cpp.md

```diff
@@ -4,8 +4,195 @@
+
+- Main module header
+- Local package headers
+- Other package headers
+- Message headers
+- Boost headers
+- C system headers
+- C++ system headers
+
+```cpp
+// Compliant
+#include "my_header.hpp"
+
+#include "my_package/foo.hpp"
+
+#include <package1/foo.hpp>
+#include <package2/bar.hpp>
+
+#include <std_msgs/msg/header.hpp>
+
+#include <iostream>
+#include <vector>
+```
+
+If you use `""` and `<>` properly, `ClangFormat` in `pre-commit` sorts headers automatically.
+
+Do not define macros between `#include` lines because it prevents automatic sorting.
+
+```cpp
+// Non-compliant
+#include <package1/foo.hpp>
+#include <package2/bar.hpp>
+
+#define EIGEN_MPL2_ONLY
+#include "my_header.hpp"
+#include "my_package/foo.hpp"
+
+#include <Eigen/Core>
+
+#include <std_msgs/msg/header.hpp>
+
+#include <iostream>
+#include <vector>
+```
+
+Instead, define macros before `#include` lines.
+
+```cpp
+// Compliant
+#define EIGEN_MPL2_ONLY
+
+#include "my_header.hpp"
+
+#include "my_package/foo.hpp"
+
+#include <Eigen/Core>
+#include <package1/foo.hpp>
+#include <package2/bar.hpp>
+
+#include <std_msgs/msg/header.hpp>
+
+#include <iostream>
+#include <vector>
+```
+
+If there are any reasons for defining macros at a specific position, write a comment before the macro.
+
+```cpp
+// Compliant
+#include "my_header.hpp"
+
+#include "my_package/foo.hpp"
+
+#include <package1/foo.hpp>
+#include <package2/bar.hpp>
+
+#include <std_msgs/msg/header.hpp>
+
+#include <iostream>
+#include <vector>
+
+// For the foo bar reason, the FOO_MACRO must be defined here.
+#define FOO_MACRO
+#include <foo/bar.hpp>
+```
+
+### Use lower snake case for function names (required, partially automated)
+
+#### Rationale
+
+- It is consistent with the C++ standard library.
+- It is consistent with other programming languages such as Python and Rust.
+
+#### Exception
+
+- For member functions of classes inherited from external project classes such as Qt, follow that naming convention.
+
+#### Reference
+
+- <https://docs.ros.org/en/humble/The-ROS2-Project/Contributing/Code-Style-Language-Versions.html#function-and-method-naming>
+
+#### Example
+
+```cpp
+void function_name()
+{
+}
+```
+
+### Use upper camel case for enum names (required, partially automated)
+
+#### Rationale
+
+- It is consistent with ROS2 libraries.
+
+#### Exception
+
+- Enums defined in the `rosidl` file can use other naming conventions.
+
+#### Reference
+
+- <http://wiki.ros.org/CppStyleGuide> (Refer to "15. Enumerations")
+
+#### Example
+
+```cpp
+enum class Color
+{
+  Red, Green, Blue
+}
+```
+
+### Use lower snake case for constant names (required, partially automated)
+
+#### Rationale
+
+- It is consistent with `std::numbers`.
+
+#### Exception
+
+- Constants defined in the `rosidl` file can use other naming conventions.
+
+#### Reference
+
+- <https://en.cppreference.com/w/cpp/numeric/constants>
+
+#### Example
+
+```cpp
+constexpr double gravity = 9.80665;
+```
+
```

Comment:
ROS2 seems to be using this convention as well [(example)](https://github.com/ros2/rclcpp/blob/33dae5d679751b603205008fcb31755986bcee1c/rclcpp/src/rclcpp/parameter_service_names.hpp#L23), I'm ok with it too.

Reply:
Thanks! Updated in 2cb800e1db8f3ce87a46d669594f9925fa1d68c2.